### PR TITLE
Online2020 homework

### PR DESCRIPTION
**Pull request by MonikaGajda**
zad 11. SFINAE  
       14. `alignas`, `alignof` 
       15. delegating constructors:
       16. lambda functions
       17. structured bindings
       18. variadic templates

### CODE REVIEW

**Review by MaciejSalamonski — CHANGES_REQUESTED** [review]
Good job +1

File: shapes/main.cpp [hunk 1]
```diff
@@ -1,89 +1,112 @@
...
     } else {
         return fibo(n - 1) + fibo(n - 2);
     }
 }
 
-int main()
-{
+template <class DerivedType, class... Arguments>
+std::shared_ptr<Shape> make_shape(Arguments&&... args) {
+    return std::shared_ptr<DerivedType>{new DerivedType{std::forward<Arguments>(args)...}};
```
Comment (ziobron) [inline shapes/main.cpp]:
```suggestion
    return std::make_shared<DerivedType>(std::forward<Arguments>(args)...);
```
your way is correct, but it's always better to allocate with make_shared

**Review by ziobron — CHANGES_REQUESTED** [review]
Delegating constructors are missing. This feature allows you to call other constructor of the same class on constructors initializer list.

File: shapes/main.cpp [hunk 2]
```diff
@@ -1,89 +1,112 @@
...
-}
+};
 
-void printCollection(const Collection& collection)
-{
-    for (const auto & it : collection)
+void printCollection(const Collection& collection) {
+    for (const auto& it : collection)
         if (it)
             it->print();
```
Comment (WBigaj) [inline shapes/main.cpp]:
```suggestion
    for (const auto& it : collection) {
        if (it) {
            it->print();
        }
    }
```

File: shapes/main.cpp [hunk 3]
```diff
@@ -1,89 +1,112 @@
...
 }
 
-void printAreas(const Collection& collection)
-{
-    for (const auto & it : collection)
+void printAreas(const Collection& collection) {
+    for (const auto& it : collection)
         if (it)
             cout << it->getArea() << std::endl;
```
Comment (WBigaj) [inline shapes/main.cpp]:
```suggestion
    for (const auto& it : collection) {
        if (it) {
            cout << it->getArea() << '\n';
        }
    }
```

**Review by WBigaj — COMMENTED** [review]
Few suggestions from me.

**Review by ziobron — CHANGES_REQUESTED** [review]
Constructor delegation is still missing. I know that a constructor with default arguments behave in the same way, but the idea behind this task was to learn constructor delegation. You need to call the constructor of the same class (not base class) on other's constructor initialization list like below:
```cpp
X::X(int a, double b, string c) {}
X::X(int a) : X(a, 2.0, "test") {}
```

**Review by ziobron — APPROVED** [review]

**Comment by ziobron** [comment]
15 XP za Nowoczesny C++ SFINAE (4447)
5 XP za Nowoczesny C++ alignas (4448)
5 XP za Nowoczesny C++ delegating constructors (4449)
5 XP za Nowoczesny C++ lambda (4450)
5 XP za Nowoczesny C++ structured bindings (4451)
15 XP za Nowoczesny C++ variadic templates (4452)

{ "requirement_ids": [4447, 4448, 4449, 4450, 4451, 4452] }

**Pull request closed without merging**